File: shared/src/commonMain/kotlin/io/mockative/krouton/AuthenticationService.kt
package io.mockative.krouton

import kotlinx.coroutines.flow.Flow
import io.mockative.krouton.Krouton
import kotlinx.coroutines.flow.StateFlow

/**
 * The authentication service
 */
@Krouton
interface AuthenticationService : RefreshableService {

    var unrelatedProperty: String

    val isLoggingIn: Flow<Boolean>

    val intFlow: Flow<Int>
    val doubleFlow: Flow<Double>

    val stateFlow: StateFlow<String?>

    // KMRoutines
    // Kallback
    //

    suspend fun login(request: AuthenticationRequest?): AuthenticationResponse

    suspend fun foo(args: List<String>)

    /**
     * Doc string
     * @param request Param String
     * @return All the flows of type [AuthenticationResponse]
     */
    fun getFlows(request: AuthenticationRequest): Flow<List<AuthenticationResponse>>

    suspend fun getFlowsAsync(request: AuthenticationRequest): Flow<List<AuthenticationResponse>>

}

File: KontinuitySample/KontinuitySample/Kontinuity/AuthenticationService.swift
import Combine
import Foundation
import shared

extension KotlinThrowable: Error {}

extension AuthenticationService {
    var isLoggingIn$: KroutonPublisher<Swift.Bool, Error> {
      KroutonPublisher { AuthenticationServiceKt.isLoggingIn(receiver: self, onElement: $0, onSuccess: $1, onFailure: $2) }
  }
    var intFlow$: KroutonPublisher<Swift.Int, Error> {
        KroutonPublisher { AuthenticationServiceKt.intFlow(receiver: self, onElement: $0, onSuccess: $1, onFailure: $2) }
    }
    
    var doubleFlow$: KroutonPublisher<Swift.Double, Error> {
        KroutonPublisher { AuthenticationServiceKt.doubleFlow(receiver: self, onElement: $0, onSuccess: $1, onFailure: $2) }
    }
    
  func login$(request: AuthenticationRequest?) -> KroutonFuture<shared.AuthenticationResponse, Error> {
      KroutonFuture { AuthenticationServiceKt.login(receiver: self, request: request, onSuccess: $0, onFailure: $1) }
  }

  func foo$(args: [String]) -> KroutonFuture<Swift.Void, Error> {
      KroutonFuture { AuthenticationServiceKt.foo(receiver: self, args: args, onSuccess: $0, onFailure: $1) }
  }

  func getFlows$(request: AuthenticationRequest) -> KroutonPublisher<[shared.AuthenticationResponse], Error> {
      KroutonPublisher { AuthenticationServiceKt.getFlows(receiver: self, request: request, onElement: $0, onSuccess: $1, onFailure: $2) }
  }
}

